### PR TITLE
[Draft] Fix Universal Compaction causing overlapping seqnos in L0 files due to overlapped seqnos between ingested files and memtable's

### PR DESCRIPTION
**Pull request by hx235**
**Context:**
Same as https://github.com/facebook/rocksdb/pull/5958#issue-511150930 but apply the fix to Universal Compaction case

Repro:
```
COERCE_CONTEXT_SWICH=1 make -j56 db_stress

./db_stress --acquire_snapshot_one_in=0 --adaptive_readahead=0 --allow_data_in_errors=True --async_io=1 --avoid_flush_during_recovery=0 --avoid_unnecessary_blocking_io=0 --backup_max_size=104857600 --backup_one_in=0 --batch_protection_bytes_per_key=0 --block_size=16384 --bloom_bits=18 --bottommost_compression_type=disable --bytes_per_sync=262144 --cache_index_and_filter_blocks=0 --cache_size=8388608 --cache_type=lru_cache --charge_compression_dictionary_building_buffer=0 --charge_file_metadata=1 --charge_filter_construction=1 --charge_table_reader=1 --checkpoint_one_in=0 --checksum_type=kCRC32c --clear_column_family_one_in=0 --compact_files_one_in=0 --compact_range_one_in=100 --compaction_pri=2 --compaction_style=1 --compaction_ttl=0 --compression_max_dict_buffer_bytes=8388607 --compression_max_dict_bytes=16384 --compression_parallel_threads=1 --compression_type=zlib --compression_use_zstd_dict_trainer=1 --compression_zstd_max_train_bytes=0 --continuous_verification_interval=0 --data_block_index_type=0 --db=/dev/shm/rocksdb_test2/rocksdb_crashtest_whitebox --db_write_buffer_size=8388608 --delpercent=4 --delrangepercent=1 --destroy_db_initially=1 --detect_filter_construct_corruption=0 --disable_wal=0 --enable_compaction_filter=0 --enable_pipelined_write=1 --fail_if_options_file_error=1 --file_checksum_impl=none --flush_one_in=1000 --format_version=5 --get_current_wal_file_one_in=0 --get_live_files_one_in=0 --get_property_one_in=0 --get_sorted_wal_files_one_in=0 --index_block_restart_interval=15 --index_type=3 --ingest_external_file_one_in=100 --initial_auto_readahead_size=0 --iterpercent=10 --key_len_percent_dist=1,30,69 --level_compaction_dynamic_level_bytes=True --log2_keys_per_lock=10 --long_running_snapshots=0 --mark_for_compaction_one_file_in=10 --max_auto_readahead_size=16384 --max_background_compactions=20 --max_bytes_for_level_base=10485760 --max_key=100000 --max_key_len=3 --max_manifest_file_size=1073741824 --max_write_batch_group_size_bytes=1048576 --max_write_buffer_number=3 --max_write_buffer_size_to_maintain=4194304 --memtable_prefix_bloom_size_ratio=0.5 --memtable_protection_bytes_per_key=1 --memtable_whole_key_filtering=1 --memtablerep=skip_list --mmap_read=1 --mock_direct_io=False --nooverwritepercent=1 --num_file_reads_for_auto_readahead=0 --num_levels=1 --open_files=100 --open_metadata_write_fault_one_in=0 --open_read_fault_one_in=32 --open_write_fault_one_in=0 --ops_per_thread=200000 --optimize_filters_for_memory=0 --paranoid_file_checks=1 --partition_filters=0 --partition_pinning=1 --pause_background_one_in=0 --periodic_compaction_seconds=0 --prefix_size=8 --prefixpercent=5 --prepopulate_block_cache=0 --progress_reports=0 --read_fault_one_in=0 --readahead_size=16384 --readpercent=45 --recycle_log_file_num=1 --reopen=20 --ribbon_starting_level=999 --snapshot_hold_ops=1000 --sst_file_manager_bytes_per_sec=0 --sst_file_manager_bytes_per_truncate=0 --subcompactions=2 --sync=0 --sync_fault_injection=0 --target_file_size_base=524288 --target_file_size_multiplier=2 --test_batches_snapshots=0 --top_level_index_pinning=3 --unpartitioned_pinning=0 --use_direct_io_for_flush_and_compaction=0 --use_direct_reads=0 --use_full_merge_v1=1 --use_merge=0 --use_multiget=1 --user_timestamp_size=0 --value_size_mult=32 --verify_checksum=1 --verify_checksum_one_in=0 --verify_db_one_in=1000 --verify_sst_unique_id_in_manifest=1 --wal_bytes_per_sync=0 --wal_compression=zstd --write_buffer_size=524288 --write_dbid_to_manifest=0 --writepercent=35 

put or merge error: Corruption: force_consistency_checks(DEBUG): VersionBuilder: L0 file #4153 with seqno 0 101850 vs. file #4164 with seqno 101187 101820
```

**Summary:**
This PR considers `earliest_mem_seqnos` whenever we are compacting to L0 level, which include the following three cases.
- [UniversalCompactionPicker::PickCompaction()](https://github.com/facebook/rocksdb/blob/7.6.fb/db/compaction/compaction_picker_universal.cc#L374) when compacting to L0 level 
     - Like the fix https://github.com/facebook/rocksdb/pull/5958#issue-511150930, we excluded files of largest seqno greater than `earliest_mem_seqno` from compaction. 
     - Firstly, this is done in `CalculateSortedRuns` at the beginning of `UniversalCompactionPicker::PickCompaction()` 
         -  However, we couldn't know the output level at this point therefore we end up filtering out such files even when we are not compacting to L0 level. 
         - Would this be problematic?? Maybe it's fine once we accumulate some data in memetable and increase the `earliest_mem_seqno` so that previously filtered out files can be chosen again.
     - Secondly, this is done in some `PickXXXCompaction()`s called in `UniversalCompactionPicker::PickCompaction()`. This is because some of these functions actually use files directly from `vstorage` as input files, not from the sanitized sorted run resulted in `CalculateSortedRuns`. This includes the following. For test coverage of each case, see **Test**.
         - `PickPeriodicCompaction()` for [multi-level](https://github.com/facebook/rocksdb/blob/7.6.fb/db/compaction/compaction_picker_universal.cc#L1276-L1277)
         - `PickCompactionToReduceSizeAmp()` for [multi-level
](https://github.com/facebook/rocksdb/blob/7.6.fb/db/compaction/compaction_picker_universal.cc#L937-L1068)         - `PickCompactionToReduceSortedRuns()` for [multi-level](https://github.com/facebook/rocksdb/blob/7.6.fb/db/compaction/compaction_picker_universal.cc#L723-L724)
         - `PickDeleteTriggeredCompaction()` for [both single and multi-level](https://github.com/facebook/rocksdb/blob/7.6.fb/db/compaction/compaction_picker_universal.cc#L1127-L1161).
             -  For this path, I also fixed a bug cased by assuming sorted_run on L0 is always a prefix of vstorage->LevelFiles(0), which is no longer true after we sanitize sorted_run in this PR.
- [CompactionPicker::CompactRange](https://github.com/facebook/rocksdb/blob/7.6.fb/db/compaction/compaction_picker.cc#L600) when compacting to L0 level
     - This include skipping such files in following functions and their derivatives: 
         - `vstorage->GetOverlappingInputs/GetOverlappingInputsRangeBinarySearch()`
         - `ExpandInputsToCleanCut()`
         - `SetupOtherInputs()`
     - However,  such treatment has failed `OutputPathId/DBTestUniversalManualCompactionOutputPathId.ManualCompactionOutputPathId/0,1` due to a [bug](https://github.com/facebook/rocksdb/pull/10812) about setting an incorrect EarliestSequenceNumber of the memtable in `DB::Reopen()`. Therefore I have to remove the reopening part in this PR https://github.com/facebook/rocksdb/pull/10776/commits/d4a2129068814a62a1ac61f1beb3b1bc47cf171f#diff-3d401dbe63b9d238f4c6ed8ae4f3b4b3b6948be976d902b7a41021b208a06192L1875-L1879 to avoid the bug and pass.
         - Such revision to the test is not ideal cuz the original test case did reflect a sequence of possible user operations. However, fixing such bug currently encounters several blockers like https://app.circleci.com/pipelines/github/facebook/rocksdb/19902/workflows/63960818-0ec0-465b-9bc8-c10ee2140816/jobs/508125 so I decided to revise the test instead in this PR.
- `CompactionPicker::CompactFiles`: fixed in https://github.com/facebook/rocksdb/pull/10777        
- Some clean-up include:
    - Made `earliest_mem_seqno` required in more functions, including those shared with level compaction in `CompactionPicker`, `DBImpl` (i.e, `DBImpl::DeleteFilesInRanges`) and `Compaction` (i.e, `Compaction::IsTrivialMove()`). For these callers `kMaxSequenceNumber` is passed in to limit this PR scope.
    - Consolidated with https://github.com/facebook/rocksdb/pull/10777's unit test under `DBCompactionTestL0FilesReorderingCorruption`
 and adopted the [comment](https://github.com/facebook/rocksdb/pull/10777#discussion_r1004882434) 
       - Also updated `TEST_SYNC_POINT` location for fifo related unit test to make it more resistant to accidental code change breaking the test.
   - Clarified comment about the basis of ordering for `VersionStorageInfo::files_`
**Test:**
- New unit tests under `DBCompactionTestL0FilesReorderingCorruption` cover all the above cases that fail before the fix and pass after, except for
   - multi-level for `UniversalCompactionPicker::PickCompaction()`, see TODO in unit test for challenges to add one
- make check
- [Ongoing] Regular CI stress run on this PR + stress test with aggressive value https://github.com/facebook/rocksdb/pull/10761 and on Universal Compaction only

### CODE REVIEW

**Comment by hx235** [comment]
Update: just a draft to see if it passes make check. Have some dependency on [Fix overlapping seqnos in L0 files cased by overlapped seqnos between ingested files and memtable (FIFO Compaction case)](https://github.com/facebook/rocksdb/pull/10777/files#top)
#10777

**Comment by hx235** [comment]
Just FYI that I am currently stress testing this PR and will let you know when there is no blocker from the signals. You can review this PR then. I also updated the PR summary to ease the review but let me know if you have any questions!

**Comment by hx235** [comment]
Update: not ready for review and just some simplification to the impl. Will go thru stress test again.

File: db/compaction/compaction_picker_universal.cc [hunk 1]
```diff
@@ -374,7 +383,7 @@ UniversalCompactionBuilder::CalculateSortedRuns(
 Compaction* UniversalCompactionBuilder::PickCompaction() {
   const int kLevel0 = 0;
   score_ = vstorage_->CompactionScore(kLevel0);
-  sorted_runs_ = CalculateSortedRuns(*vstorage_);
+  sorted_runs_ = CalculateSortedRuns(*vstorage_, earliest_mem_seqno_);
```
Comment (hx235) [inline db/compaction/compaction_picker_universal.cc:386]: As mentioned in the PR summary, we couldn't know the output level at this point therefore we might end up filtering out such files even when we are not compacting to L0 level and the filtered files are harmless.

Would this be problematic?? Maybe it's fine once we accumulate some data in memetable and increase the earliest_mem_seqno so that previously filtered out files can be chosen again.

**Comment by hx235** [comment]
@ajkr Feel free to postpone reviewing this PR (i.e, the newest two comments - not too much change actually) till after https://github.com/facebook/rocksdb/pull/10777 is landed and this one is rebased (see PR summary for more).

File: db/compaction/compaction_picker_universal.cc (hunk 1 above)
Comment (hx235) [reply db/compaction/compaction_picker_universal.cc:386]: Actually I can look into whether multi-level universal compaction has something like IntraL0Compaction(). If there isn't such thing, then we can assume multi-level universal compaction always compact to Ln>0 and single-level always compact to L0. Downside is we are making this assumption that's subtle and might not get updated if we ever want to have IntraL0Compaction for multi-level universal.

**Review by ajkr — COMMENTED** [review]
I think we can't simply `return nullptr` because that indicates to `RunManualCompaction()` that the compaction completed, while actually it never ran. The previous commit (skipping L0 files when they overlap memtable seqno) looks right, though.

One big assumption that the previous commit makes is that universal compaction always flushes the memtable. Otherwise, there could be cases where a file is ingested before `CompactRange()` is called, that file's keys overlap with the compacted range, but that file is skipped over during compaction due to its largest_seqno overlapping a memtable. I think that assumption holds today.

File: HISTORY.md [hunk 2]
```diff
@@ -26,6 +26,7 @@
 * Fixed a memory safety bug in experimental HyperClockCache (#10768)
 * Fixed some cases where `ldb update_manifest` and `ldb unsafe_remove_sst_file` are not usable because they were requiring the DB files to match the existing manifest state (before updating the manifest to match a desired state).
 * Fix FIFO compaction causing corruption of overlapping seqnos in L0 files due to ingesting files of overlapping seqnos with memtable's under `CompactionOptionsFIFO::allow_compaction=true` or `CompactionOptionsFIFO::age_for_warm>0` or `CompactRange()/CompactFiles()` is used. Before the fix, `force_consistency_checks=true` may catch the corruption before it's exposed to readers, in which case writes returning `Status::Corruption` would be expected.
+* Fix Universal compaction causing corruption of overlapping seqnos in L0 files due to ingesting files of overlapping seqnos with memtable's, including the case when `CompactRange()/CompactFiles()` is used. Before the fix, `force_consistency_checks=true` may catch the corruption before it's exposed to readers, in which case writes returning `Status::Corruption` would be expected.
```
Comment (hx235) [inline HISTORY.md:29]: @ajkr for the case where CompactRange (and CompactFile ?) can compact to non-L0 then move to L0, is there a specific option to do that? I want to note it in HISTORY that this case is current NOT fixed.

**Comment by hx235** [comment]
Not ready for review - more subpaths of `UniversalCompactionPicker::PickCompact()` needs similar treatment and UT coverage. And stress test is yet to be run.

**Comment by hx235** [comment]
> One big assumption that the previous commit makes is that universal compaction always flushes the memtable.

File: db/db_compaction_test.cc [hunk 3]
```diff
@@ -6464,82 +6458,366 @@ TEST_P(DBCompactionTestWithParam,
...
+
+class DBCompactionTestL0FilesReorderingCorruptionWithParam
+    : public DBCompactionTestL0FilesReorderingCorruption,
+      public testing::WithParamInterface<CompactionStyle> {
+ public:
+  DBCompactionTestL0FilesReorderingCorruptionWithParam()
+      : DBCompactionTestL0FilesReorderingCorruption() {}
+};
+
+// TODO: add `CompactionStyle::kCompactionStyleLevel` to
```
Comment (hx235) [inline db/db_compaction_test.cc:6808]: I haven't investigated but when num_level  = 1 with level compaction, there will still be a L1 in vstorage and ingested file will be placed in L1, creating difficulties to force compaction to L0. So I skipped adding UT for kCompactionStyleLevel. However let me know if you still see value in investigating this and figure this out!

File: db/db_compaction_test.cc [hunk 4]
```diff
@@ -6464,82 +6458,366 @@ TEST_P(DBCompactionTestWithParam,
...
+  DBCompactionTestL0FilesReorderingCorruption() : DBCompactionTest() {}
+  void SetupOptions(const CompactionStyle compaciton_style,
+                    const std::string compaction_path_to_test = "") {
+    options_ = CurrentOptions();
+    options_.create_if_missing = true;
+    options_.compression = kNoCompression;
+
+    options_.force_consistency_checks = true;
+    options_.compaction_style = compaciton_style;
+    // TODO: expand test coverage to num_lvels > 1 for universal compacion,
```
Comment (hx235) [inline db/db_compaction_test.cc:6472]: I wanted to expand test coverage to num_levels > 1 for universal compaction as mentioned in PR summary to cover some edge case in PickCompaction() but couldn't quite force the compaction to L0 - any good idea?

File: db/db_compaction_test.cc [hunk 5]
```diff
@@ -6464,82 +6458,366 @@ TEST_P(DBCompactionTestWithParam,
...
+ private:
+  const Snapshot* snapshot_ = nullptr;
+  std::atomic<bool> compaction_path_sync_point_called_;
+};
+
+TEST_F(DBCompactionTestL0FilesReorderingCorruption,
+       FlushAfterIntraL0UniversalCompactionWithIngestedFile) {
+  for (const std::string compaction_path_to_test :
+       {"PickPeriodicCompaction", "PickCompactionToReduceSizeAmp",
+        "PickCompactionToReduceSortedRuns"}) {
```
Comment (hx235) [inline db/db_compaction_test.cc:6660]: I tried to cover `PickDeleteTriggeredCompaction()` but only come up with low-level test case manipulating vstorage directly like https://github.com/facebook/rocksdb/blob/main/db/compaction/compaction_picker_test.cc#L2953-L2977. Is there another way of doing this? If not, I can still go with the low-level one - better than nothing.

**Comment by hx235** [comment]
Running stress test now - will notify @ajkr once signals are clear

File: db/db_compaction_test.cc (hunk 3 above)
Comment (ajkr) [reply db/db_compaction_test.cc:6808]: I think we sanitize num_levels to be at least 3 or something in level compaction.

File: db/compaction/compaction_picker.cc [hunk 6]
```diff
@@ -656,7 +671,9 @@ Compaction* CompactionPicker::CompactRange(
       inputs[level - start_level].level = level;
       auto& files = inputs[level - start_level].files;
       for (FileMetaData* f : vstorage->LevelFiles(level)) {
-        files.push_back(f);
+        if (output_level > 0 || f->fd.largest_seqno <= earliest_mem_seqno) {
```
[the remark this answers is not in the document]
Comment (hx235) [reply db/compaction/compaction_picker.cc:674]: oh okay - that seems like a contradiction between CompactRange() contract (i.e, "a key should be compacted .... ") and the "skip file" approach we are pursuing here .... Have we discussed the solution to this before? 

Maybe leave as TODO for CompactionPicker ::CompactRange() for now and fix it by not allowing such "L0 file overlapping memtable in seqno" exists as you mentioned in the task about next step for universal compaction?

File: db/compaction/compaction_picker_universal.cc [hunk 7]
```diff
@@ -340,17 +346,26 @@ void UniversalCompactionBuilder::SortedRun::DumpSizeInfo(
 
 std::vector<UniversalCompactionBuilder::SortedRun>
 UniversalCompactionBuilder::CalculateSortedRuns(
-    const VersionStorageInfo& vstorage) {
+    const VersionStorageInfo& vstorage,
+    const SequenceNumber earliest_mem_seqno) {
   std::vector<UniversalCompactionBuilder::SortedRun> ret;
   for (FileMetaData* f : vstorage.LevelFiles(0)) {
+    if (f->fd.largest_seqno > earliest_mem_seqno) {
```
[the remark this answers is not in the document]
Comment (hx235) [reply db/compaction/compaction_picker_universal.cc:353]: Aha good point ... I didn't even realize there is a feature called level0_stop_writes_trigger ... need to think about this more. Current situation sounds like a deadlock to me

File: HISTORY.md [hunk 8]
```diff
@@ -26,6 +26,7 @@
 * Fixed a memory safety bug in experimental HyperClockCache (#10768)
 * Fixed some cases where `ldb update_manifest` and `ldb unsafe_remove_sst_file` are not usable because they were requiring the DB files to match the existing manifest state (before updating the manifest to match a desired state).
 * Fix FIFO compaction causing corruption of overlapping seqnos in L0 files due to ingesting files of overlapping seqnos with memtable's under `CompactionOptionsFIFO::allow_compaction=true` or `CompactionOptionsFIFO::age_for_warm>0` or `CompactRange()/CompactFiles()` is used. Before the fix, `force_consistency_checks=true` may catch the corruption before it's exposed to readers, in which case writes returning `Status::Corruption` would be expected.
```
[the remark this answers is not in the document]
Comment (hx235) [reply HISTORY.md:28]: Fixed now.

File: HISTORY.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (hx235) [reply HISTORY.md]: Fixed

Comment (hx235) [reply HISTORY.md]: Thanks - will specify this situation not being fixed in HISTORY for now.

Comment (hx235) [reply HISTORY.md]: And it seems weird to me that the ReFitLevel https://github.com/facebook/rocksdb/blob/main/db/db_impl/db_impl_compaction_flush.cc#L1204 happens after we compact range with the output level computed **before** but not **after** considering change_level. So like the output_level we use in compact range is not final final.

Comment (hx235) [reply HISTORY.md]: Updated the HISTORY.md to mention this.

File: db/db_compaction_test.cc (hunk 3 above)
Comment (hx235) [reply db/db_compaction_test.cc:6808]: Oh okay thanks! Then CompactRangeOptions::change_level=true and CompactRangeOptions::target_level=0 might help with the unit test design. Will try

File: db/compaction/compaction_picker_universal.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (hx235) [reply db/compaction/compaction_picker_universal.cc:353]: A side-note is that db stress can be improved to detect deadlock like this.

File: db/db_compaction_test.cc (hunk 5 above)
Comment (hx235) [reply db/db_compaction_test.cc:6660]: Actually - found another more useful test case TEST_P(DBCompactionTestWithParam, CompactionDeletionTrigger)

File: db/compaction/compaction_picker_universal.cc [hunk 9]
```diff
@@ -1150,8 +1205,13 @@ Compaction* UniversalCompactionBuilder::PickDeleteTriggeredCompaction() {
       if (sr->being_compacted) {
         continue;
       }
-      FileMetaData* f = vstorage_->LevelFiles(0)[loop];
+      FileMetaData* f = sr->file;
```
Comment (hx235) [inline db/compaction/compaction_picker_universal.cc:1208]: [Note] A bug fix - cased by assuming sorted_run on L0 is always a prefix of vstorage->LevelFiles(0), which is no longer true after we sanitize sorted_run in this PR.

File: db/compaction/compaction_picker_universal.cc [hunk 10]
```diff
@@ -1169,7 +1229,12 @@ Compaction* UniversalCompactionBuilder::PickDeleteTriggeredCompaction() {
         break;
       }
 
-      FileMetaData* f = vstorage_->LevelFiles(0)[loop];
+      FileMetaData* f = sr->file;
```
Comment (hx235) [inline db/compaction/compaction_picker_universal.cc:1232]: [Note] A bug fix - cased by assuming sorted_run on L0 is always a prefix of vstorage->LevelFiles(0), which is no longer true after we sanitize sorted_run in this PR.

File: db/db_compaction_test.cc (hunk 5 above)
Comment (hx235) [reply db/db_compaction_test.cc]: Fixed

**Comment by hx235** [comment]
Update: addressed some comments + added more UT + fixed a bug along the way

File: db/compaction/compaction_picker.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (ajkr) [reply db/compaction/compaction_picker.cc:674]: No solution has been discussed before, but a necessary condition for this bug ("If the memtable were not flushed, ...") is not present in universal CompactRange(), because it always does the flush. My original point was just that this is a big assumption that might change or might not even be true if I misread the code.

**Comment by ajkr** [comment]
What if we set FileMetaData::largest_seqno as follows?

- During flush, set it to the max of (i) the max seqno of keys in the new file and (ii) one higher than the max largest_seqno of all L0 files. Today it currently is set to (i).
- During ingestion, set it as we do today
- During intra-L0, set it to the max largest_seqno of all input files

**Comment by ajkr** [comment]
> What if we set FileMetaData::largest_seqno as follows?
>
> - During flush/ingestion, set it to the max of (i) the max seqno of keys in the new file and (ii) one higher than the max largest_seqno of all L0 files. Today it currently is set to (i).
> - During intra-L0, set it to the max largest_seqno of all input files

The idea is to play with the metadata so we don't have to worry about what files can or cannot be picked for compaction. I'm still working on an intuitive explanation for this altered meaning of largest_seqno. In case we can't find one, I also thought about a new ordering field in T133335138.

**Comment by ajkr** [comment]
> * During intra-L0, set it to the max largest_seqno of all input files

HBase does this -- https://github.com/apache/hbase/blob/984d226010558c5b5e54d283b465a645816ffb24/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/compactions/Compactor.java#L171-L172. But ever since HBase style compaction was introduced to RocksDB in https://github.com/facebook/rocksdb/commit/554c06dd18d32866eb2430497a70c947923a4dbb, we have been learning largest_seqno from the output file keys rather than from the input file metadata. While our way is buggy when ordering L0 files by descending largest_seqno, it still feels like we have a more intuitive meaning of largest_seqno, so probably we should introduce a different field like "L0 epoch number" to use for ordering.

**Comment by hx235** [comment]
Closed since alternative approach was pursued.
